### PR TITLE
fix for multi user scenario

### PR DESCRIPTION
**Pull request by xieofxie**
They are not thread-safe when multiple users call these functions.

Also we must use `services.AddSingleton<IBot, TestBot>()` for the state of RandomSelector._rand

### CODE REVIEW

**Comment by boydc2014** [comment]
Looks good to me, Random.Next is not thread-safe.  Request Tom and ChrisMc to review and confirm. 

@xieofxie  do you have a scenario that multi-thread is used, maybe we can check in with some tests if it's a typical use case?

**Comment by xieofxie** [comment]
@boydc2014 Unfortunately, after some searching, I didn't find examples about how to write tests for concurrency.

I write one myself here https://github.com/xieofxie/botbuilder-dotnet/commit/79c86cc756bc15a17392f15f000bda8fc0edc2a4, but it looks not good enough for me.

Typically this happens when bot is used by multiple users and they happen to run into same functions.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Selectors/RandomSelector.cs [hunk 1]
```diff
@@ -79,8 +79,11 @@ public Task<IReadOnlyList<int>> Select(SequenceContext context, CancellationToke
             var result = new List<int>();
             if (candidates.Count > 0)
             {
-                var selection = _rand.Next(candidates.Count);
-                result.Add(candidates[selection]);
+                lock (this)
+                {
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Selectors/RandomSelector.cs]: You technically only need the lock around the _rand.Next.  The result is local.

Comment (xieofxie) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Selectors/RandomSelector.cs]: Right, changed~

**Pull request merged**